src/app/common/TrayView.vue

```diff
@@ -0,0 +1,64 @@
+<template>
+  <KSlideout
+    close-button-alignment="end"
+    :has-overlay="false"
+    is-visible
+    prevent-close-on-blur
+    data-testid="tray"
+    @close="emit('close')"
+  >
+    <template
+      v-if="$slots.icon"
+      #before-title
+    >
+      <div class="tray-title-wrapper">
+        <slot name="icon" />
+
+        <h2 class="tray-title">
+          <slot name="title" />
+        </h2>
+      </div>
```

Comment:
Why are we putting the title in the "before-title", I get that we can't use `:title` here but can't we just put it in the default slot? I think generally its also a bad idea to put `h*` tags. With reusable components you never know what the level of headers where you are using the component.

Reply:
Separate to this, why are we only showing the title if there is an icon. Could we just put all this in the default slot please?

Reply:
> Why are we putting the title in the "before-title", I get that we can't use `:title` here but can't we just put it in the default slot?

Because that’s not where the title belongs. The KSlideout title should ideally go into a `title` slot. In the absence of such a slot (and the inadequacy of `props.title`), we put the title in the `before-title` slot. That’s not ideal, but something we’ll be able to move away from with changes to KSlideout.

> I think generally its also a bad idea to put h* tags.

In absence of a more flexible pattern, that’s what we have to do. It’s also what we do virtually everywhere in our application. We could provide the level/tag via prop, but I don’t think that’s something we should consider here as part of this work.

> Separate to this, why are we only showing the title if there is an icon. Could we just put all this in the default slot please?

That’s a mistake which I’ve fixed now. The condition was meant to be on the icon slot.

Reply:
>  It’s also what we do virtually everywhere in our application. 

Not sure how factually correct that is ^, but lets not get into that just yet.

> In absence of a more flexible pattern

Putting an AppView inside the default slot and reusing its header/title is a more flexible pattern. This is exactly the same as we do with KCard everywhere in our application. KCard has a header slot but we never use it.

Reply:
Just to be clear, unless things change what I would like to happen here is that:

1. We should avoid using `h*` headers is non-route components.
2. We should use AppView#title for the header, AppView should go in the default slot. The argument about "keeping things consistent with the design" by using the component is not valid as the font styles for the KSlideout title are dependant on you using the `:title` attribute, not the `before-title` slot. There is no difference in "keeping consistency" with using the before-title slot or the default slot. If anything its worse to use the before-title slot as you have solving something with a hack that is probably unexpected by the author of the component, seeing as in this case we have a very simple alternative, we should use that alternative at least in the short term.

Honestly the more I look at this and the more I look at it and think about it, the less happy I am with the approach taken here. Especially considering how straight forwards the alternative I've been proposing is.

Reply:
> Putting an AppView inside the default slot and reusing its header/title is a more flexible pattern.

Not very flexible if it prevents the use of titles in things like title slots, is it?

> 1. We should avoid using `h*` headers in non-route components.

That’s a bad rule. Headings need to appear where they need to appear and if that’s in a non-view component it’s in a non-view component.

> The argument about "keeping things consistent with the design" by using the component is not valid as the font styles for the KSlideout title are dependant on you using the `:title` attribute, not the `before-title` slot.

So you agree that if we had access to a title slot (proposed in https://github.com/Kong/kongponents/pull/1770), it would be a valid argument?

> There is no difference in "keeping consistency" with using the before-title slot or the default slot. If anything its worse to use the before-title slot as you are solving something with a hack that is probably unexpected by the author of the component.

You know for a fact that the only reason I’m using the hacky way of populating the title using the `before-title` slot is because there is no `title` slot.

> Honestly the more I look at this and the more I look at it and think about it, the less happy I am with the approach taken here. Especially considering how straight forwards the alternative I've been proposing is.

I have changed this now to your liking, but I will re-visit using KSlideout’s title mechanisms once they have improved. It’s already gotten more problematic to get the alignment of the close button with the content right.

Reply:
> That’s a bad rule. Headings need to appear where they need to appear and if that’s in a non-view component it’s in a non-view component.

We should use a slot so we can specify the level of the heading from The Outside.

> So you agree that if we had access to a title slot (proposed in https://github.com/Kong/kongponents/pull/1770), it would be a valid argument?

I think I saw that it was suggested that those PRs should be closed because other people are working on the component, my comment there suggests the same.

As to whether its a valid argument. I don't know until I see the redesigned/rebuilt component, it might end up being completely different to what we have now. What I do know is, it will probably use the shared styles from the design tokens, which we use everywhere already for our headers, and I also know that we use KCard with headers in its 'content' slot even though they provide a title slot. I should also note the `:title` is an optional attribute.

Once we have a new KSlideout, we can look to see if we can integrate further with it (something which I'm not against doing). In the meantime we aren't confusing matters by adding the title using the before-title slot, even though we don't get any of the designed styles by doing that.

> You know for a fact that the only reason I’m using the hacky way of populating the title using the before-title slot is because there is no title slot.

Yes I know, I get why you are doing it, thats why I was saying if we can't use ':title' we may as well use the approach to headers we already have i.e. AppView#title, and in not using `before-title` we aren't doing something unexpected to the author of the component.

> I have changed this now to your liking, but I will re-visit using KSlideout’s title mechanisms once they have improved. It’s already gotten more problematic to get the alignment of the close button with the content right.

Awesome thankyou! Much appreciated, let me take another look at this now.

Maybe take a look at my commit again, I took extra care there to make sure the baselines where lined up.

![Screenshot 2023-10-20 at 12 39 21](https://github.com/kumahq/kuma-gui/assets/554604/59ef3be0-b43e-4d97-a5e9-710804aca545)